### PR TITLE
laser: implement codecopy and extcodecopy

### LINKED ISSUE #269: Implement extcodecopy
## Description
Currently the codecopy instruction for laser  is not implemented

### PR DESCRIPTION
**Pull request by kielbarry**
This is in reference to Issue #269 (marked as good first issue). I am 100% sure this is not 100% correct (I will be reaching out on gitter). So please review extremely critically. The code I have added follows the structure set by the completed CALLDATACOPY opcode.

Edit( by @JoranHonig): fixes #269

### CODE REVIEW

**Comment by b-mueller** [comment]
> I am 100% sure this is not 100% correct 
 
🤣 

Sure, let's talk on Gitter :)

File: mythril/laser/ethereum/svm.py [hunk 1]
```diff
@@ -671,19 +671,97 @@ def _sym_exec(self, gblState):
...
                 addr = state.stack.pop()
                 start, s2, size = state.stack.pop(), state.stack.pop(), state.stack.pop()
 
+                try:
+                    mstart = helper.get_concrete_int(start)
+                except:
+                    logging.debug("Unsupported symbolic memory offset in EXTCODECOPY")
+                    continue
+                try:
+                    dstart = helper.get_concrete_int(s1)
```
Comment (norhh) [inline mythril/laser/ethereum/svm.py]: should be s2 right?

Comment (kielbarry) [reply mythril/laser/ethereum/svm.py]: thank you for all your catches!

File: mythril/laser/ethereum/svm.py [hunk 2]
```diff
@@ -671,19 +671,97 @@ def _sym_exec(self, gblState):
...
+                except:
+                    logging.debug("Unsupported symbolic code offset in EXTCODECOPY")
+                    state.mem_extend(mstart, 1)
+                    state.memory[mstart] = BitVec("code_" + str(addr) + "_cpy", 256)
+                    continue
+                try:
+                    size = helper.get_concrete_int(op2)
```
[the remark this answers is not in the document]
Comment (kielbarry) [reply mythril/laser/ethereum/svm.py]: I think so, thanks for the catch.

File: mythril/laser/ethereum/svm.py [hunk 3]
```diff
@@ -671,19 +671,97 @@ def _sym_exec(self, gblState):
...
+                    continue
+                try:
+                    dstart = helper.get_concrete_int(s1)
+                except:
+                    logging.debug("Unsupported symbolic code offset in CODECOPY")
+                    state.mem_extend(mstart, 1)
+                    state.memory[mstart] = BitVec("code_" + str(disassembly.instruction_list) + "_cpy", 256)
+                    continue
+                try:
+                    size = helper.get_concrete_int(op2)
```
[the remark this answers is not in the document]
Comment (kielbarry) [reply mythril/laser/ethereum/svm.py]: I think so, thanks for the catch.

File: mythril/laser/ethereum/svm.py [hunk 4]
```diff
@@ -671,19 +671,97 @@ def _sym_exec(self, gblState):
...
+                    try:
+                        state.mem_extend(mstart, size)
+                    except:
+                        logging.debug("Memory allocation error: mstart = " + str(mstart) + ", size = " + str(size))
+                        state.mem_extend(mstart, 1)
+                        state.memory[mstart] = BitVec("code_" + str(disassembly.instruction_list) + "_" + str(dstart), 256)
+                        continue
+
+                    try:
+                        i_data = disassembly.instruction_list[dstart]
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/laser/ethereum/svm.py]: Hi @kielbarry, now its easy to get bytecode from dissassembly as bytecode variable is added to disassembly.

File: mythril/laser/ethereum/svm.py [hunk 5]
```diff
@@ -671,19 +671,97 @@ def _sym_exec(self, gblState):
...
+                    try:
+                        state.mem_extend(mstart, size)
+                    except:
+                        logging.debug("Memory allocation error: mstart = " + str(mstart) + ", size = " + str(size))
+                        state.mem_extend(mstart, 1)
+                        state.memory[mstart] = BitVec("extcode_" + str(addr) + "_" + str(dstart), 256)
+                        continue
+
+                    try:
+                        state.memory[i] = addr
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/laser/ethereum/svm.py]: hi @kielbarry, if you are not sure how to get the code externally into the memory then you can refer to the extcodesize function in instructions.py. Also as worklist pr is merged all the instructions are in the instructions.py currently

**Comment by JoranHonig** [comment]
Hi @kielbarry, this pr is getting a bit old now. I'm working on #191 and that requires these ops to be implemented.
I'm going to take your implemenentation for CODECOPY and finalize it, so I can continue

**Comment by b-mueller** [comment]
I assume this is superseded by #379?

**Comment by JoranHonig** [comment]
I'll close this pr now to keep oversight, @kielbarry if you feel like starting again on this please do :+1:

**Pull request closed without merging**